cpp/src/arrow/util/io_util.cc

```diff
@@ -1867,5 +1889,42 @@ uint64_t GetOptionalThreadId() {
   return (tid == 0) ? tid - 1 : tid;
 }
 
+// Returns the current resident set size (physical memory use) measured
+// in bytes, or zero if the value cannot be determined on this OS.
+// See: https://stackoverflow.com/a/14927379
```

Comment:
In general we try to avoid directly copy/pasting chunks of code from SO for licensing reasons but I don't know that we ever got an authoritative answer for this.  Either way, I think things are paraphrased enough by this point it isn't a concern.